### PR TITLE
#399 Singlefile support

### PR DESCRIPTION
**Pull request by cdvv7788**
# Summary

Add https://github.com/gildas-lormeau/SingleFile as an extractor

**Related issues: #399 

# Changes these areas

- [ ] Bugfixes
- [X] Feature behavior
- [ ] Command line interface
- [ ] Configuration options
- [ ] Internal architecture
- [ ] Archived data layout on disk

### CODE REVIEW

**Comment by cdvv7788** [comment]
@pirate This first version is functional but it is lacking a lot of flexibility. It uses the `CHROME_BINARY`. I need to add new options to be able to use other browsers.
Also, the installation is not possible via package managers, like it is for `wget` and `chrome`. I had to clone it, install npm dependencies and create a symlink to my `bin` folder. Should we leave this method enabled by default?

It currently works, if you have chrome installed. It creates a new file in the archive named `single-file.html`. Currently, it is created directly by the `single-file` process.

File: archivebox/config/__init__.py [hunk 1]
```diff
@@ -104,13 +105,15 @@
...
         'USE_WGET':                 {'type': bool,  'default': True},
+        'USE_SINGLEFILE':           {'type': bool,  'default': True},
         'USE_GIT':                  {'type': bool,  'default': True},
         'USE_CHROME':               {'type': bool,  'default': True},
         'USE_YOUTUBEDL':            {'type': bool,  'default': True},
 
         'CURL_BINARY':              {'type': str,   'default': 'curl'},
         'GIT_BINARY':               {'type': str,   'default': 'git'},
         'WGET_BINARY':              {'type': str,   'default': 'wget'},
+        'SINGLEFILE_BINARY':        {'type': str, 'default': 'single-file'},
```
Comment (pirate) [inline archivebox/config/__init__.py:116]:
```suggestion
        'SINGLEFILE_BINARY':        {'type': str,   'default': 'single-file'},
```

File: archivebox/extractors/singlefile.py [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+
+
+@enforce_types
+def save_singlefile(link: Link, out_dir: Optional[str]=None, timeout: int=TIMEOUT) -> ArchiveResult:
+    """download full site using single-file"""
+
+    out_dir = out_dir or link.link_dir
+    output = str(Path(out_dir).absolute() / "single-file.html")
+
+    # WGET CLI Docs: https://www.gnu.org/software/wget/manual/wget.html
```
Comment (pirate) [inline archivebox/extractors/singlefile.py:39]:
```suggestion
    # SingleFile CLI Docs: https://github.com/gildas-lormeau/SingleFile/tree/master/cli
```

File: archivebox/extractors/singlefile.py [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+    SINGLEFILE_BINARY,
+    SINGLEFILE_VERSION,
+    CHROME_BINARY,
+)
+from ..logging_util import TimedProgress
+
+
+@enforce_types
+def should_save_singlefile(link: Link, out_dir: Optional[str]=None) -> bool:
+    out_dir = out_dir or link.link_dir
```
Comment (pirate) [inline archivebox/extractors/singlefile.py:25]:
```suggestion
    output = Path(out_dir).absolute() / "single-file.html"
```

File: archivebox/extractors/singlefile.py [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+    if not os.path.exists(out_dir):
+        return False
+
+    return SAVE_SINGLEFILE
```
Comment (pirate) [inline archivebox/extractors/singlefile.py:29]:
```suggestion
    output = Path(out_dir or link.link_dir) / 'single-file.html'
    return SAVE_SINGLEFILE and (not output.exists())
```

**Comment by pirate** [comment]
A good start, things left:

- add an iframe to the link_details.html template for singlefile
- add a files column icon for SingleFile in the core.admin snapshot list
- add it to `archviebox.index.latest_outputs`
- add it to `archviebox.index.canonoical_outputs`
- anywhere else you can fine (search for `dom_path` and add it next to it everywhere)

**Comment by cdvv7788** [comment]
- [ ] add an iframe to the link_details.html template for singlefile
- [ ] add a files column icon for SingleFile in the core.admin snapshot list
- [ ] add it to archviebox.index.latest_outputs
- [ ] add it to archviebox.index.canonoical_outputs
- [ ] anywhere else you can fine (search for dom_path and add it next to it everywhere)

**Comment by cdvv7788** [comment]
@pirate please review. I think we can close this and add the new config options to the cli in another PR. Let me know if you prefer adding that in here.

File: archivebox/extractors/singlefile.py [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+    output = Path(out_dir or link.link_dir) / 'single-file.html'
+    return SAVE_SINGLEFILE and (not output.exists())
+
+
+@enforce_types
+def save_singlefile(link: Link, out_dir: Optional[str]=None, timeout: int=TIMEOUT) -> ArchiveResult:
+    """download full site using single-file"""
+
+    out_dir = out_dir or link.link_dir
+    output = str(Path(out_dir).absolute() / "single-file.html")
```
Comment (pirate) [inline archivebox/extractors/singlefile.py:35]:
```suggestion
    output = str(Path(out_dir).absolute() / "singlefile.html")
```

File: archivebox/extractors/singlefile.py [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+    CHROME_BINARY,
+)
+from ..logging_util import TimedProgress
+
+
+@enforce_types
+def should_save_singlefile(link: Link, out_dir: Optional[str]=None) -> bool:
+    out_dir = out_dir or link.link_dir
+
+    output = Path(out_dir or link.link_dir) / 'single-file.html'
```
Comment (pirate) [inline archivebox/extractors/singlefile.py:26]:
```suggestion
    output = Path(out_dir or link.link_dir) / 'singlefile.html'
```

File: archivebox/index/schema.py [hunk 7]
```diff
@@ -365,6 +365,7 @@ def is_archived(self) -> bool:
             'screenshot.png',
             'output.html',
             'media',
+            'single-file.html'
```
Comment (pirate) [inline archivebox/index/schema.py:368]:
```suggestion
            'singlefile.html'
```

File: archivebox/index/schema.py [hunk 8]
```diff
@@ -406,6 +406,7 @@ def canonical_outputs(self) -> Dict[str, Optional[str]]:
             'google_favicon_path': 'https://www.google.com/s2/favicons?domain={}'.format(self.domain),
             'wget_path': wget_output_path(self),
             'warc_path': 'warc',
+            'singlefile_path': 'single-file.html',
```
Comment (pirate) [inline archivebox/index/schema.py:409]:
```suggestion
            'singlefile_path': 'singlefile.html',
```

File: archivebox/themes/legacy/link_details.html [hunk 9]
```diff
@@ -335,6 +336,18 @@ <h5>🗃 Files</h5>
...
                     </div>
+                    <div class="col-lg-2">
+                        <div class="card">
+                          <iframe class="card-img-top" src="$singlefile_path" sandbox="allow-same-origin allow-scripts allow-forms" scrolling="no"></iframe>
+                          <div class="card-body">
+                            <a href="$singlefile_path" style="float:right" title="Open in new tab..." target="_blank" rel="noopener">
+                                <img src="../../static/external.png" class="external"/>
+                            </a>
+                            <a href="$singlefile_path" target="preview"><h4 class="card-title">SingleFile</h4></a>
+                            <p class="card-text">archive/single-file.html</p>
```
Comment (pirate) [inline archivebox/themes/legacy/link_details.html:347]:
```suggestion
                            <p class="card-text">archive/singlefile.html</p>
```

File: tests/test_extractors.py [hunk 10]
```diff
@@ -2,4 +2,10 @@
...
 def test_wget_broken_pipe(tmp_path, process):
     add_process = subprocess.run(['archivebox', 'add', 'http://127.0.0.1:8080/static/example.com.html'], capture_output=True)
-    assert "TypeError chmod_file(..., path: str) got unexpected NoneType argument path=None" not in add_process.stdout.decode("utf-8")
\ No newline at end of file
+    assert "TypeError chmod_file(..., path: str) got unexpected NoneType argument path=None" not in add_process.stdout.decode("utf-8")
+
+def test_singlefile_works(tmp_path, process):
+    add_process = subprocess.run(['archivebox', 'add', 'http://127.0.0.1:8080/static/example.com.html'], capture_output=True) 
+    archived_item_path = list(tmp_path.glob('archive/**/*'))[0]
+    output_file = archived_item_path / "single-file.html"
```
Comment (pirate) [inline tests/test_extractors.py:10]:
```suggestion
    output_file = archived_item_path / "singlefile.html" 
```

**Comment by pirate** [comment]
The last thing remaining is to add it to the docker image so it works out-of-the-box.

**Comment by cdvv7788** [comment]
@pirate Added a fixture to the tests to avoid using methods we were not actually testing. By default it will bypass all the extractors. The python tests should be running faster now because of this change.
I also added the singlefile installation as part of the process (it was that or remove the singlefile test).
Additionally, I added a config variable that was missing (`SAVE_SINGLEFILE`) and that was causing erratic behaviour.

**Pull request merged**